### PR TITLE
Add protocol (e.g. http or https) to ServerConfiguration

### PR DESCRIPTION
**Pull request by kristofhennebel**
Add protocol (e.g. http or https) to ServerConfiguration

### CODE REVIEW

File: single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java [hunk 1]
```diff
@@ -59,7 +63,8 @@ public ServerConfiguration(
             long connectionHealthCheckPeriodInSeconds,
             long healthReadTimeoutInMilliseconds,
             int failureResponseCountThreshold,
-            Consumer<HttpRequest.Builder> requestTransformer
+            Consumer<HttpRequest.Builder> requestTransformer,
+            String protocol
```
Comment (nhenneaux) [inline single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java:67]: can you add a new constructor without protocol to avoid breaking changes and revert the change in the tests? Then add another test with http protocol?

File: single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java [hunk 2]
```diff
@@ -69,6 +74,7 @@ public ServerConfiguration(
         this.healthReadTimeoutInMilliseconds = healthReadTimeoutInMilliseconds;
         this.failureResponseCountThreshold = failureResponseCountThreshold;
         this.requestTransformer = requestTransformer;
+        this.protocol = protocol;
```
Comment (nhenneaux) [inline single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java:77]: I would validate the value of the protocol either http or https since nothing else is expected and throw an illegalargumentexception if invalid

**Comment by nhenneaux** [comment]
https://nicolas.henneaux.io is unreachable for now (I have some issue with OCA that deleted my setup), could you remove the endpoint from the list of endpoints?

File: single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java (hunk 1 above)
Comment (kristofhennebel) [reply single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java:83]: ah yes, good one, done

File: single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java (hunk 2 above)
Comment (kristofhennebel) [reply single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java:96]: yep, done

File: monitored-httpclient/src/test/java/com/github/nhenneaux/resilienthttpclient/monitoredclientpool/HttpClientPoolTest.java [hunk 3]
```diff
@@ -831,7 +823,7 @@ void shouldUpdateToFailedCountForHealthChecksFailed() {
     void shouldDecommissionIfCouldNotFulfillFailedResponseCountThresholdRequirement() throws IOException, URISyntaxException, InterruptedException {
         // Given
         final String hostname = "postman-echo.com";
-        final ServerConfiguration serverConfiguration = new ServerConfiguration(hostname, -1, "", 1, 1, -1, 0, DEFAULT_REQUEST_TRANSFORMER);
+        final ServerConfiguration serverConfiguration = new ServerConfiguration(hostname, -1, "", 1, 1, -1, 0, DEFAULT_REQUEST_TRANSFORMER, "https");
```
Comment (nhenneaux) [inline monitored-httpclient/src/test/java/com/github/nhenneaux/resilienthttpclient/monitoredclientpool/HttpClientPoolTest.java:826]: you can revert those changes in tests since now it is https by default?

File: single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java (hunk 2 above)
Comment (nhenneaux) [reply single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java:96]: very minor - `PROTOCOLS` could be names SUPPORTED_PROTOCOLS and the error message could use the list instead of re-enumerating them

File: single-host-httpclient/src/test/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfigurationTest.java [hunk 4]
```diff
@@ -58,7 +61,7 @@ void shouldProperlyReturnConfiguredValues() {
         Assertions.assertEquals(444L, serverConfiguration.getDnsLookupRefreshPeriodInSeconds());
         Assertions.assertEquals(111L, serverConfiguration.getHealthReadTimeoutInMilliseconds());
         Assertions.assertEquals(1234, serverConfiguration.getPort());
```
Comment (nhenneaux) [inline single-host-httpclient/src/test/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfigurationTest.java:63]:
```suggestion
        Assertions.assertEquals("http", serverConfiguration.getProtocol());
```

**Comment by nhenneaux** [comment]
> https://nicolas.henneaux.io is unreachable for now (I have some issue with OCA that deleted my setup), could you remove the endpoint from the list of endpoints?

I finally got it fixed using Cloudflare, could you re-active it. Very sorry for changing again :(

**Comment by nhenneaux** [comment]
I fixed nicolas.henneaux.io tests failing and java23 issue in https://github.com/nhenneaux/resilient-httpclient/pull/67, can you rebase on master?

File: monitored-httpclient/src/test/java/com/github/nhenneaux/resilienthttpclient/monitoredclientpool/HttpClientPoolTest.java (hunk 3 above)
Comment (kristofhennebel) [reply monitored-httpclient/src/test/java/com/github/nhenneaux/resilienthttpclient/monitoredclientpool/HttpClientPoolTest.java]: done

File: single-host-httpclient/src/test/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfigurationTest.java (hunk 4 above)
Comment (kristofhennebel) [reply single-host-httpclient/src/test/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfigurationTest.java:63]: done

File: single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java [hunk 5]
```diff
@@ -1,18 +1,21 @@
...
-    private static final int DEFAULT_FAILURE_RESPONSE_COUNT_THRESHOLD = -1; // It means no validation by failed response count
+    public static final int DEFAULT_PORT = -1;
+    public static final String DEFAULT_HEALTH_PATH = "";
+    public static final long DEFAULT_DNS_LOOKUP_REFRESH_PERIOD_IN_SECONDS = TimeUnit.MINUTES.toSeconds(5);
+    public static final long DEFAULT_CONNECTION_HEALTH_CHECK_PERIOD_IN_SECONDS = 30;
+    public static final long DEFAULT_HEALTH_READ_TIMEOUT_IN_MILLISECONDS = TimeUnit.SECONDS.toMillis(5);
+    public static final int DEFAULT_FAILURE_RESPONSE_COUNT_THRESHOLD = -1; // It means no validation by failed response count
     public static final Consumer<HttpRequest.Builder> DEFAULT_REQUEST_TRANSFORMER = null;
+    public static final String DEFAULT_PROTOCOL = "https";
+    public static final List<String> PROTOCOLS = List.of("http", "https");
```
Comment (nhenneaux) [inline single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java:18]: minor - It can be a Set since always used to verify contains

**Comment by nhenneaux** [comment]
two tests are failing
```
Error:    ResilientClientTest.throwForInvalidUrl:125 Unexpected exception type thrown, expected: <java.lang.IllegalArgumentException> but was: <java.lang.NullPointerException>
Error:    SingleIpHttpClientTest.shouldFailOnMalformedUrl:150 expected: <Cannot build health URI from ServerConfiguration{hostname='com.github.nhenneaux.resilienthttpclient.monitoredclientpool.SingleIpHttpClientTest.shouldCreateClientWithoutRefresh', port=-234, healthPath='&dfsfsd', connectionHealthCheckPeriodInSeconds=1, dnsLookupRefreshPeriodInSeconds=1, healthReadTimeoutInMilliseconds=-1, failureResponseCountThreshold= 0}> but was: <Cannot build health URI from ServerConfiguration{hostname='com.github.nhenneaux.resilienthttpclient.monitoredclientpool.SingleIpHttpClientTest.shouldCreateClientWithoutRefresh', port=-234, healthPath='&dfsfsd', connectionHealthCheckPeriodInSeconds=1, dnsLookupRefreshPeriodInSeconds=1, healthReadTimeoutInMilliseconds=-1, failureResponseCountThreshold= 0, protocol= https}>

File: single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java (hunk 2 above)
Comment (kristofhennebel) [reply single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java:96]: yep, done

File: single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java (hunk 5 above)
Comment (kristofhennebel) [reply single-host-httpclient/src/main/java/com/github/nhenneaux/resilienthttpclient/singlehostclient/ServerConfiguration.java]: yep, done

**Comment by nhenneaux** [comment]
Another test failing with Set instead of list

Supported protocols are [http, https], but was: abc> but was: <Supported protocols are [https, http], but was: abc>

**Comment by nhenneaux** [comment]
I have just deployed v1.0.4, it will be very soon available in maven central https://repo1.maven.org/maven2/com/github/nhenneaux/resilienthttpclient/resilient-httpclient/1.0.4/
